client/src/connection/rest/common.ts

```diff
@@ -0,0 +1,132 @@
+import * as typer from "media-typer";
+import { Link } from "./api/compute";
+import { Configuration } from "./api/configuration";
+
+import { AxiosRequestConfig, AxiosResponse } from "axios";
+import { createRequestFunction, RequestArgs } from "./api/common";
+
+/**
+ * States that can be on a Job or Session
+ */
+export enum ComputeState {
+  Error = "error",
+  Canceled = "canceled",
+  Done = "done",
+  Warning = "warning",
+  Completed = "completed",
+  Running = "running",
+  Pending = "pending",
+  Idle = "idle",
+}
+
+export interface stateOptions {
+  onChange?: boolean;
+  wait?: number;
+}
+
+export interface BaseCompute {
+  links?: Array<Link>;
+}
+
+//global api config
+const apiConfig = new Configuration({ baseOptions: {} });
+
+export function getApiConfig(): Configuration {
+  return apiConfig;
+}
+
+export function computeMediaType(type: string): string {
+  const parsed = typer.parse(type);
+  if (
+    parsed.type === "application" &&
+    parsed.subtype?.startsWith("vnd.sas") &&
+    parsed.suffix === undefined
+  ) {
+    parsed.suffix = "json";
+    return typer.format(parsed);
+  }
+  return type;
+}
+
+/**
+ * Base class for compute like objects
+ */
+export class Compute {
+  etag: string;
+
+  async self<T>(): Promise<T> {
+    throw new Error("Not implemented");
+  }
+
+  //Shortcut function to get a link by its name
+  getLink(links: Array<Link>, rel: string): Link | undefined {
+    return links.find((x) => x.rel === rel);
+  }
+
+  /*
+  Get the options for a link
+  This sets the URL, accept/content-type headers as well as the Method and 
+  etag headers if needed
+  */
+  getLinkOptions(link: Link, options?: AxiosRequestConfig): RequestArgs {
+    if (options === undefined) options = {};
+    const headers = { ...options?.headers };
+
+    if (link.method === "POST") {
+      if (link.type !== undefined)
+        headers["Content-Type"] = computeMediaType(link.type);
+      if (link.responseType !== undefined)
+        headers["Accept"] = computeMediaType(link.responseType);
+    } else if (link.method === "PUT") {
+      if (link.type !== undefined)
+        headers["Content-Type"] = computeMediaType(link.type);
+      if (link.responseType !== undefined)
+        headers["Accept"] = computeMediaType(link.responseType);
+      if (this.etag !== undefined) headers["If-Match"] = this.etag;
+    } else if (link.method === "DELETE") {
+      if (this.etag !== undefined) headers["If-Match"] = this.etag;
+    } else if (link.type !== undefined) {
+      headers["Accept"] = computeMediaType(link.type);
+    }
+
+    //The link must have an href
+    if (!link.href) throw new Error();
+
+    /*
+    const url = new URL(`http://${link.href}`);
+    const parms = {};
+    url.searchParams.forEach((value: string, key: string) => {
+      parms[key] = value;
+    });
+    */
```

Comment:
Is this needed?

Reply:
No, I removed it.

Reply:
There was a problem with a new version of Axios where query parms on the URL were getting dropped. This was my workaround, but the problem was fixed in Axios, so if we upgrade to the latest version it should work.